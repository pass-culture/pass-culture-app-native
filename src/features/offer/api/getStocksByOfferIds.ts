import { api } from 'api/api'
import { ApiError } from 'api/ApiError'
import { OfferNotFound } from 'features/offer/pages/OfferNotFound/OfferNotFound'
import { OfferNotFoundError } from 'libs/monitoring'
import { LogTypeEnum } from 'libs/monitoring/errors'

export async function getStocksByOfferIds(offerIds: number[], logType: LogTypeEnum) {
  if (offerIds.length === 0) {
    return { offers: [] }
  }
  try {
    return await api.postNativeV2OffersStocks({
      offer_ids: offerIds,
    })
  } catch (error) {
    if (error instanceof ApiError && error.statusCode === 404) {
      // This happens when the offer has been rejected but it is still indexed on Algolia
      // due to asynchronous reindexing of the back office
      throw new OfferNotFoundError(offerIds, {
        Screen: OfferNotFound,
        logType,
      })
    }
    throw error
  }
}
