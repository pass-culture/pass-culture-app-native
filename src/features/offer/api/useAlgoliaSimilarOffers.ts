import { useMemo } from 'react'
import { useQuery } from 'react-query'

import { getSimilarOffersInOrder } from 'features/offer/helpers/getSimilarOffersInOrder/getSimilarOffersInOrder'
import { useIsUserUnderage } from 'features/profile/helpers/useIsUserUnderage'
import { IncompleteSearchHit } from 'libs/algolia'
import { fetchOffersByIds } from 'libs/algolia/fetchAlgolia/fetchOffersByIds'
import { useTransformOfferHits, filterOfferHit } from 'libs/algolia/fetchAlgolia/transformOfferHit'
import { QueryKeys } from 'libs/queryKeys'
import { Offer } from 'shared/offer/types'

export const useAlgoliaSimilarOffers = (
  ids: string[],
  shouldPreserveIdsOrder?: boolean
): Offer[] | undefined => {
  const isUserUnderage = useIsUserUnderage()
  const transformHits = useTransformOfferHits()

  const { data: hits } = useQuery(
    [QueryKeys.SIMILAR_OFFERS, JSON.stringify(ids)],
    () => fetchOffersByIds({ objectIds: ids, isUserUnderage }),
    { enabled: ids.length > 0 }
  )

  return useMemo(() => {
    if (!hits || hits.length === 0) return

    if (shouldPreserveIdsOrder) {
      const offers = getSimilarOffersInOrder(ids, hits)
      return (offers as IncompleteSearchHit[]).filter(filterOfferHit).map(transformHits) as Offer[]
    }

    return (hits as IncompleteSearchHit[]).filter(filterOfferHit).map(transformHits) as Offer[]
  }, [hits, ids, shouldPreserveIdsOrder, transformHits])
}
