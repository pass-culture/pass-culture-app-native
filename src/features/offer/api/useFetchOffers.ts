import { useQuery } from 'react-query'

import { fetchOffers } from 'libs/algolia/fetchAlgolia/fetchOffers'
import { QueryKeys } from 'libs/queryKeys'

export const useFetchOffers = (...params: Parameters<typeof fetchOffers>) => {
  return useQuery({
    queryKey: [QueryKeys.SEARCH_RESULTS, params],
    queryFn: () => fetchOffers(...params),
  })
}
