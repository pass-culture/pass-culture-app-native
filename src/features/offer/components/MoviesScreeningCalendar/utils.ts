import { MOVIE_CALENDAR_PADDING } from 'features/offer/components/MovieCalendar/MovieCalendar'

export const handleMovieCalendarScroll = (
  currentIndex: number,
  flatListWidth: number,
  itemWidth: number
) => {
  const shift = flatListWidth / 2 - MOVIE_CALENDAR_PADDING
  const centerOfSelectedElement = currentIndex * itemWidth + itemWidth / 2

  if (centerOfSelectedElement - shift < 0 || currentIndex === 0) {
    return { offset: 0 }
  }
  return { offset: centerOfSelectedElement - shift }
}
