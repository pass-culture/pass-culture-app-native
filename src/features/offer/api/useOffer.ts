import { useQuery } from 'react-query'

import { api } from 'api/api'
import { ApiError } from 'api/apiHelpers'
import { OfferResponse } from 'api/gen'
import { OfferNotFound } from 'features/offer/pages/OfferNotFound'
import { OfferNotFoundError } from 'libs/monitoring'
import { useNetwork } from 'libs/network/useNetwork'
import { QueryKeys } from 'libs/queryKeys'

async function getOfferById(offerId: number) {
  if (!offerId) {
    throw new OfferNotFoundError(offerId, OfferNotFound)
  }
  try {
    return await api.getnativev1offerofferId(offerId)
  } catch (error) {
    if (error instanceof ApiError && error.statusCode === 404) {
      throw new OfferNotFoundError(offerId, OfferNotFound)
    }
    throw error
  }
}

export const useOffer = ({ offerId }: { offerId: number }) => {
  const { isConnected } = useNetwork()

  return useQuery<OfferResponse | undefined>(
    [QueryKeys.OFFER, offerId],
    () => (offerId ? getOfferById(offerId) : undefined),
    { enabled: isConnected }
  )
}
