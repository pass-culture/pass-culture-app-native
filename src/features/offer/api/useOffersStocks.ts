import { useQuery } from 'react-query'

import { api } from 'api/api'
import { ApiError } from 'api/ApiError'
import { OffersStocksResponseV2 } from 'api/gen'
import { OfferNotFound } from 'features/offer/pages/OfferNotFound/OfferNotFound'
import { useRemoteConfigContext } from 'libs/firebase/remoteConfig'
import { OfferNotFoundError } from 'libs/monitoring'
import { useNetInfoContext } from 'libs/network/NetInfoWrapper'
import { QueryKeys } from 'libs/queryKeys'

async function getStocksByOfferIds(offerIds: number[], shouldLogInfo: boolean) {
  if (offerIds.length === 0) {
    throw new OfferNotFoundError(offerIds, {
      Screen: OfferNotFound,
      shouldBeCapturedAsInfo: shouldLogInfo,
    })
  }
  try {
    // eslint-disable-next-line @typescript-eslint/return-await
    return await api.postNativeV2OffersStocks({
      offer_ids: offerIds,
    })
  } catch (error) {
    if (error instanceof ApiError && error.statusCode === 404) {
      // This happens when the offer has been rejected but it is still indexed on Algolia
      // due to asynchronous reindexing of the back office
      throw new OfferNotFoundError(offerIds, {
        Screen: OfferNotFound,
        shouldBeCapturedAsInfo: shouldLogInfo,
      })
    }
    throw error
  }
}

export const useOffersStocks = ({ offerIds }: { offerIds: number[] }) => {
  const netInfo = useNetInfoContext()
  const { shouldLogInfo } = useRemoteConfigContext()

  return useQuery<OffersStocksResponseV2>(
    [QueryKeys.OFFER, offerIds],
    () => getStocksByOfferIds(offerIds, shouldLogInfo),
    { enabled: !!netInfo.isConnected }
  )
}
