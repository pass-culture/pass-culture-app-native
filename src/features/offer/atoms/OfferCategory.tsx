import { t } from '@lingui/macro'
import React from 'react'

import { CategoryIdEnum } from 'api/gen'
import { mapCategoryToIcon } from 'libs/parsers'

import { IconWithCaption } from './IconWithCaption'

interface OfferCategoryProps {
  categoryId: CategoryIdEnum | null
  label: string
}

export const OfferCategory = ({ categoryId, label }: OfferCategoryProps) => {
  const Icon = mapCategoryToIcon(categoryId)
  return (
    <IconWithCaption Icon={Icon} caption={label} accessibilityLabel={t`CatÃ©gorie de l'offre`} />
  )
}
