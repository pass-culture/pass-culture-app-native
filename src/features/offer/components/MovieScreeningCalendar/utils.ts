import { OfferStockResponse } from 'api/gen'

export const getDateString = (date: string) => new Date(date).toDateString()

export const sortScreeningDates = (screeningDates: OfferStockResponse[]) =>
  [...screeningDates].sort((a, b) => {
    return a.beginningDatetime && b.beginningDatetime
      ? new Date(a.beginningDatetime).getTime() - new Date(b.beginningDatetime).getTime()
      : 0
  })
