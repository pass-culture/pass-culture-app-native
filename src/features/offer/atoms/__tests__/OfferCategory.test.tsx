import React from 'react'

import { CategoryIdEnum } from 'api/gen'
import { render } from 'tests/utils'

import { OfferCategory } from '../OfferCategory'

describe('OfferCategory', () => {
  it('renders correctly', () => {
    const { toJSON } = render(
      <OfferCategory categoryId={CategoryIdEnum.LIVRE} label="Livre ou carte lecture" />
    )
    expect(toJSON()).toMatchSnapshot()
  })
})
