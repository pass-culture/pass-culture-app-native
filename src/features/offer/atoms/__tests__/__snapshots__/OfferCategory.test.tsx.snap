// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferCategory renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "paddingBottom": 4,
          "paddingLeft": 4,
          "paddingRight": 4,
          "paddingTop": 4,
        },
      ]
    }
  >
    <View
      fill="#626262"
      height={40}
      width={40}
    >
      <Text>
        undefined-SVG-Mock
      </Text>
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#151515",
          "fontFamily": "Montserrat-SemiBold",
          "fontSize": 12,
          "lineHeight": 16,
          "textAlign": "center",
        },
      ]
    }
    testID="caption-undefined"
  >
    Livre ou carte lecture
  </Text>
</View>
`;
